### ISSUE TITLE
currentEventsByTag query loops forever when in LookingForDelayed (backtracking) state

### ISSUE THREAD

**Issue by wketting** [akka/akka-persistence-cassandra#206]
I've a local setup with a rather large event database. Not always, but in my current setup 90% of the time after a rather large (and fast) replay, when at the end looking for delayed events is entered, but it does never come out of it, forever repeating the last backtracking query, as this log shows:
```
...
 DEBUG a.p.c.q.EventsByTagPublisher - query chunk done for tag [ar-user], timBucket [20170503], count [18] of [18]
 DEBUG a.p.c.q.EventsByTagPublisher - query for tag [ar-user] from [20170504] [dc33cd30-2fe8-11e7-9a13-db82eebc2bd5 (2017-05-03 10:11:23)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - query chunk done for tag [ar-user], timBucket [20170504], count [0] of [0]
 DEBUG a.p.c.q.EventsByTagPublisher - query for tag [ar-user] from [20170505] [dc33cd30-2fe8-11e7-9a13-db82eebc2bd5 (2017-05-03 10:11:23)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - query chunk done for tag [ar-user], timBucket [20170505], count [17] of [17]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query for tag [ar-user] from [20170505] [35abe720-315c-11e7-8080-808080808080 (2017-05-05 06:29:36)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query chunk done for tag [ar-user], timBucket [20170505], count [0] of [1]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query for tag [ar-user] from [20170505] [35abe720-315c-11e7-8080-808080808080 (2017-05-05 06:29:36)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query chunk done for tag [ar-user], timBucket [20170505], count [0] of [1]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query for tag [ar-user] from [20170505] [35abe720-315c-11e7-8080-808080808080 (2017-05-05 06:29:36)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query chunk done for tag [ar-user], timBucket [20170505], count [0] of [1]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query for tag [ar-user] from [20170505] [35abe720-315c-11e7-8080-808080808080 (2017-05-05 06:29:36)] limit [500]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query chunk done for tag [ar-user], timBucket [20170505], count [0] of [1]
 DEBUG a.p.c.q.EventsByTagPublisher - backtracking delayed query for tag [ar-user] from [20170505] [35abe720-315c-11e7-8080-808080808080 (2017-05-05 06:29:36)] limit [500]
...
```

What I think is happening is the following: The backtracking query gets a resultset with 1 entry (last entry of today == last timebucket), however this event has already been processed, so is marked as duplicate. Therefore line EventsByTagPublisher:271 will never become true:
```
replayCountZero = retrievedCount == 0
```
Because retrievedCount is 1 (the duplicate event found). This blocks the stopIfDone check in line 219 of the same file:
```
def stopIfDone(): Unit = {
    if (buf.isEmpty && (isToOffsetDone || (replayCountZero && isCurrentTimeAfterToOffset()))) {
      onCompleteThenStop()
    }
  }
```
So the loop will never stop.

The reason this sometimes works (in my setup) is that backtracking sometimes is started before the last timebucket has been reached. This happens when the next timebucket is already started but not finished yet. I think that is caused by setting currTimeBucket to the last timeBucket before the previous is finished. But this is another issue...

I think that a simple fix can be: Extend ReplayDone case class with duplicateCount property which is increased in EventsByTagsFetcher:151 (case SequenceNumbers.Before =>), and change the check at EventsByTagPublisher:271 into:
```
      replayCountZero = (retrievedCount - duplicateCount) == 0
```
I'm not too familiair with this code, so I'm not sure if I miss anything with the proposed solution.

**Comment by patriknw** [comment]
Thanks for reporting, I'll look into this next week. Just to confirm, what version are you using?

It would be very helpful If you are able to create a reproducer in this test: https://github.com/akka/akka-persistence-cassandra/blob/master/core/src/test/scala/akka/persistence/cassandra/query/EventsByTagSpec.scala#L823
It doesn't have to fail, but illustrate the scenario with debug logging turned on.

**Comment by wketting** [comment]
Sorry, forgot to mention: I'm using 0.52 (latest). I'll try to create reproducer the coming day(s).

**Comment by patriknw** [comment]
thanks

**Comment by wketting** [comment]
It is really difficult to reproduce the behavior: Sofar I am not able to reproduce the exact behavior. This test: https://github.com/akka/akka-persistence-cassandra/compare/master...wketting:wip-currenteventsbytag-loop?expand=1 is coming close:

1. Start replay: currentEventsByTag
2.Make sure after reaching the current timebucket backtracking is activated. 
3. During backtracking query at least 1 event should be found, but ignored since it's a duplicate.
4. (Now the test continuous with a normal query of the same current timebucket. After the query the test stops.)
During my local situation (with real cassandra etc.) in step 4 The test continuous with the same backtracking query (backtracking mode is immediately reactivated or still active?). This situation loops forever. 

I'm not sure if the events found in previous timebuckets could influence the flow. Getting in the loop seems only to occur with specific timings?

Shouldn't currentEventsByTag be stopped after the last backtracking query of the current timebucket? If no duplicate event is found, it is stopped correctly (I think).

**Comment by patriknw** [comment]
thanks, I'll take a look now

**Comment by patriknw** [comment]
@wketting I have created a reproducer and a fix. Would be great if you can review/verify. Let's continue discussion in my PR https://github.com/akka/akka-persistence-cassandra/pull/207

**Comment by wketting** [comment]
Great! Gonna look at right away!

**Comment by patriknw** [comment]
@wketting released 0.28 and 0.53 with this fixed

**Comment by wketting** [comment]
Thanks a lot!
